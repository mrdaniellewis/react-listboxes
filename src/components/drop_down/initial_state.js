export function initialState({ selectedOption }) {
  return {
    expanded: false,
    listClassName: null,
    listStyle: null,
    search: '',
    focusedOption: selectedOption,
  };
}
