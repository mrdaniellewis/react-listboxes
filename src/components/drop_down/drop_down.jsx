import React, { useRef, useEffect, useLayoutEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { useThunkReducer as useReducer } from '../../hooks/use_thunk_reducer.js';
import { reducer } from './reducer.js';
import { initialState } from './initial_state.js';
import {
  clearSearch, onKeyDown, setFocusedIndex, onBlur,
  onToggleOpen, onFocus, onButtonKeyDown, onClick,
  onOptionsChanged,
} from './actions.js';
import { Context } from '../../context.js';
import { options as validateOptions } from '../../validators/options.js';
import { useNormalisedOptions } from '../../hooks/use_normalised_options.js';
import { useOnBlur } from '../../hooks/use_on_blur.js';
import { usePrevious } from '../../hooks/use_previous.js';
import { componentCustomiser } from '../../validators/component_customiser.js';
import { renderGroupedOptions } from '../../helpers/render_grouped_options.js';
import { dismemberComponent } from '../../helpers/dismember_component.js';
import { classGenerator } from '../../helpers/class_generator.js';

export function DropDown(rawProps) {
  const optionisedProps = useNormalisedOptions(rawProps, { mustHaveSelection: true });
  const {
    options, value: _, setValue, id, className,
    children, selectedIndex, managedFocus,
    ButtonComponent, ListBoxComponent, OptionComponent,
    GroupComponent, ValueComponent, DropDownComponent,
    ...componentProps
  } = optionisedProps;
  const buttonRef = useRef();
  const listRef = useRef();
  const focusedRef = useRef();

  const [state, dispatch] = useReducer(
    reducer,
    { ...optionisedProps, buttonRef, listRef },
    initialState,
    id,
  );
  const { expanded, search, focusedIndex } = state;
  const onBlurHandler = useOnBlur(() => dispatch(onBlur()), listRef);

  const prevOptions = usePrevious(options);
  useEffect(() => (
    dispatch(onOptionsChanged(prevOptions))
  ), [prevOptions]);

  useEffect(() => {
    if (!search || !expanded) {
      return undefined;
    }
    const found = options.findIndex((o) => (
      !o.unselectable && o.label.toLowerCase().startsWith(search)
    ));
    if (found > -1) {
      dispatch(setFocusedIndex(found));
    }
    const timeout = setTimeout(() => dispatch(clearSearch()), 1000);

    return () => clearTimeout(timeout);
  }, [options, search, expanded, setValue]);

  useLayoutEffect(() => {
    if (expanded && options[focusedIndex] && managedFocus) {
      focusedRef.current.focus();
    } else if (expanded) {
      listRef.current.focus();
    }
  }, [expanded, managedFocus, options, focusedIndex]);

  const customDropDownComponent = dismemberComponent(DropDownComponent, 'div');
  const customButtonComponent = dismemberComponent(ButtonComponent, 'button');
  const customListBoxComponent = dismemberComponent(ListBoxComponent, 'ul');
  const customGroupComponent = dismemberComponent(GroupComponent, 'li');
  const customOptionComponent = dismemberComponent(OptionComponent, 'li');
  const customValueComponent = dismemberComponent(ValueComponent);

  const classes = classGenerator(className);

  return (
    <Context.Provider value={{ dispatch, ...optionisedProps, listRef, buttonRef, ...state }}>
      <customDropDownComponent.type
        {...(customDropDownComponent.type === Fragment ? undefined : { className })}
        {...customDropDownComponent.props}
        {...componentProps}
      >
        <customButtonComponent.type
          type="button"
          id={id}
          aria-controls={`${id}_listbox`}
          aria-expanded={expanded ? 'true' : null}
          aria-haspopup="listbox"
          ref={buttonRef}
          onClick={() => dispatch(onToggleOpen())}
          onKeyDown={(e) => dispatch(onButtonKeyDown(e))}
          className={classes('button')}
          {...customButtonComponent.props}
        >
          {children ?? options[selectedIndex].label}
        </customButtonComponent.type>
        <customListBoxComponent.type
          ref={listRef}
          id={`${id}_listbox`}
          role="listbox"
          hidden={!expanded}
          tabIndex={-1}
          aria-activedescendant={options[focusedIndex]?.key || null}
          onFocus={(e) => dispatch(onFocus(e))}
          onBlur={onBlurHandler}
          onKeyDown={(e) => dispatch(onKeyDown(e))}
          className={classes('listbox')}
          {...customListBoxComponent.props}
        >
          {renderGroupedOptions({
            options,
            renderGroup(group) {
              const { key, html, label, index, children: groupChildren } = group;
              return (
                <Context.Provider
                  key={key}
                  value={{ dispatch, ...optionisedProps, ...state, group }}
                >
                  <customGroupComponent.type
                    id={key}
                    ref={index === focusedIndex ? focusedRef : null}
                    data-focused={index === focusedIndex ? 'true' : null}
                    role="group"
                    aria-label={label}
                    tabIndex={-1}
                    aria-disabled="true"
                    className={classes('listbox__group', index === focusedIndex && '--focused')}
                    {...customGroupComponent.props}
                    {...html}
                  >
                    {label}
                  </customGroupComponent.type>
                  {groupChildren}
                </Context.Provider>
              );
            },
            // eslint-disable-next-line react/prop-types
            renderOption(option) {
              const { label, key, html, disabled, index, selected, group } = option;
              return (
                <Context.Provider
                  key={key}
                  value={{ dispatch, ...optionisedProps, ...state, option }}
                >
                  <customOptionComponent.type
                    id={key}
                    role="option"
                    tabIndex={-1}
                    aria-selected={selected ? 'true' : null}
                    aria-disabled={disabled ? 'true' : null}
                    ref={index === focusedIndex ? focusedRef : null}
                    className={classes('listbox__option', index === focusedIndex && '--focused', group && '--grouped')}
                    {...customOptionComponent.props}
                    {...html}
                    onClick={disabled ? null : (e) => dispatch(onClick(e, option))}
                  >
                    <customValueComponent.type
                      {...customValueComponent.props}
                    >
                      {label}
                    </customValueComponent.type>
                  </customOptionComponent.type>
                </Context.Provider>
              );
            },
          })}
        </customListBoxComponent.type>
      </customDropDownComponent.type>
    </Context.Provider>
  );
}

DropDown.propTypes = {
  blank: PropTypes.string,
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  options: validateOptions.isRequired,
  setValue: PropTypes.func.isRequired,
  value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
  managedFocus: PropTypes.bool,
  className: PropTypes.string,
  ListBoxComponent: componentCustomiser,
  ButtonComponent: componentCustomiser,
  GroupComponent: componentCustomiser,
  OptionComponent: componentCustomiser,
  ValueComponent: componentCustomiser,
  DropDownComponent: componentCustomiser,
};

DropDown.defaultProps = {
  blank: '',
  children: null,
  value: null,
  className: 'dropdown',
  managedFocus: true,
  ListBoxComponent: null,
  ButtonComponent: null,
  GroupComponent: null,
  OptionComponent: null,
  ValueComponent: null,
  DropDownComponent: null,
};
