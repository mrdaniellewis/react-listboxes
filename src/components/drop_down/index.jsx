import React, { useRef, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { PopupButton } from '../popup_button.jsx';
import { ListBox } from '../list_box.jsx';
import { useThunkReducer as useReducer } from '../../hooks/use_thunk_reducer.js';
import { reducer } from './reducer.js';
import { initialState } from './initial_state.js';
import { clearSearch, setExpanded, onKeyDown } from './actions.js';
import { Context } from '../../context.js';
import { options as validateOptions } from '../../validators/options.js';
import { optionise } from '../../helpers/optionise.js';

export function DropDown(props) {
  const { children, id, options: rawOptions, setValue, value } = props;
  const options = useMemo(() => rawOptions.map(optionise), [rawOptions]);
  const [state, dispatch] = useReducer(reducer, props, initialState, id);
  const { expanded, search } = state;
  const currentOption = options.find(({ value: optionValue }) => optionValue === value);

  const buttonRef = useRef();
  const listRef = useRef();

  function onBlur() {
    setTimeout(() => {
      if (!listRef.current.contains(document.activeElement)) {
        dispatch(setExpanded(false));
        buttonRef.current.focus();
      }
    }, 0);
  }

  useEffect(() => {
    const found = options.find(option => option.label.startsWith(search));
    setValue(found.value);
    // TODO: Need to debouce this
    setTimeout(() => dispatch(clearSearch()), 1000);
  }, [options, search, setValue]);

  return (
    <Context.Provider value={{ ...props, expanded, options, setExpanded }}>
      <PopupButton
        ref={buttonRef}
        targetRef={listRef}
        hasPopup="listbox"
        setExpanded={newExpanded => dispatch(setExpanded(newExpanded))}
      >
        {children || currentOption ? currentOption.label : ''}
      </PopupButton>
      <ListBox
        id={id}
        options={options}
        hidden={!expanded}
        ref={listRef}
        onKeyDown={e => dispatch(onKeyDown(e))}
        onBlur={onBlur}
        setValue={newValue => setValue(newValue)}
        value={value}
      />
    </Context.Provider>
  );
}

DropDown.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  options: validateOptions.isRequired,
  setValue: PropTypes.func.isRequired,
  value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
};

DropDown.defaultProps = {
  children: null,
  value: null,
};
