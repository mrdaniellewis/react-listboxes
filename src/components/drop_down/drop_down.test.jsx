import React, { useState } from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { DropDown } from './drop_down.jsx';

function DropDownWrapper({ value: initialValue, ...props }) {
  const [value, setValue] = useState(initialValue);
  return (
    <DropDown id="id" value={value} setValue={setValue} {...props} />
  );
}

describe('options', () => {
  describe('as array of strings', () => {
    const options = ['Apple', 'Banana', 'Orange'];

    it('renders a closed drop down', () => {
      const { container } = render(<DropDownWrapper options={options} />);
      expect(container).toMatchSnapshot();
      expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
    });

    it('renders a drop down with a selected value', () => {
      const { container } = render(<DropDownWrapper options={options} value="Orange" />);
      const button = container.querySelector('button');
      expect(button).toHaveTextContent('Orange');
    });

    describe('expanding the list box', () => {
      it('opens the drop down on click with the first option selected', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        expect(container.querySelector('[role=listbox]')).not.toHaveAttribute('hidden');
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });

      it('opens the drop down on click with the value selected', () => {
        const { container } = render(<DropDownWrapper options={options} value="Orange" />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        expect(container.querySelector('[role=listbox]')).not.toHaveAttribute('hidden');
        expect(document.activeElement).toEqual(container.querySelector('[role=option]:nth-child(3)'));
      });

      it('opens the drop down with the down arrow', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.keyDown(button, { key: 'ArrowDown' });
        expect(container.querySelector('[role=listbox]')).not.toHaveAttribute('hidden');
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });

      it('opens the drop down with the up arrow', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.keyDown(button, { key: 'ArrowUp' });
        expect(container.querySelector('[role=listbox]')).not.toHaveAttribute('hidden');
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });
    });

    describe('navigating options', () => {
      it('moves to the next option with the down arrow', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]:nth-child(2)'));
      });

      it('moves to the first option from the last option with the down arrow', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
        fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
        fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });

      it('moves to the previous option with the up arrow', () => {
        const { container } = render(<DropDownWrapper options={options} value="Banana" />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'ArrowUp' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });

      it('moves to the last option from the first option with the up arrow', () => {
        const { container } = render(<DropDownWrapper options={options} />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'ArrowUp' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]:nth-child(3)'));
      });

      it('moves to the first option with the home key', () => {
        const { container } = render(<DropDownWrapper options={options} value="Banana" />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'Home' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]'));
      });

      it('moves to the last option with the end key', () => {
        const { container } = render(<DropDownWrapper options={options} value="Banana" />);
        const button = container.querySelector('button');
        fireEvent.click(button);
        fireEvent.keyDown(document.activeElement, { key: 'End' });
        expect(document.activeElement).toEqual(container.querySelector('[role=option]:nth-child(3)'));
      });

      it.todo('moves the option when typing');
      it.todo('moves the option when typing multiple letters');
      it.todo('clears typing after a short delay');
    });

    describe('selecting an option', () => {
      describe('when clicking on an option', () => {
        describe('when the value is not the current value', () => {
          it('calls setValue', () => {
            const spy = jest.fn();
            const { container } = render(<DropDownWrapper options={options} setValue={spy} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.click(container.querySelector('[role=option]:nth-child(2)'));
            expect(spy).toHaveBeenCalledWith('Banana');
          });

          it('closes the list box and selects the button', () => {
            const { container } = render(<DropDownWrapper options={options} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.click(container.querySelector('[role=option]:nth-child(2)'));
            expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            expect(document.activeElement).toEqual(button);
          });

          it('updates the displayed value', () => {
            const { container } = render(<DropDownWrapper options={options} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.click(container.querySelector('[role=option]:nth-child(2)'));
            expect(button).toHaveTextContent('Banana');
          });
        });

        describe('when the value is the current value', () => {
          it('closes the listbox without calling setValue', () => {
            const spy = jest.fn();
            const { container } = render(<DropDownWrapper options={options} value="Banana" setValue={spy} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.click(container.querySelector('[role=option]:nth-child(2)'));
            expect(spy).not.toHaveBeenCalled();
            expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            expect(document.activeElement).toEqual(button);
          });
        });
      });

      describe('when pressing enter on an option', () => {
        describe('when the value is not the current value', () => {
          it('calls setValue', () => {
            const spy = jest.fn();
            const { container } = render(<DropDownWrapper options={options} setValue={spy} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            fireEvent.keyDown(document.activeElement, { key: 'Enter' });
            expect(spy).toHaveBeenCalledWith('Banana');
          });

          it('closes the list box and selects the button', () => {
            const { container } = render(<DropDownWrapper options={options} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            fireEvent.keyDown(document.activeElement, { key: 'Enter' });
            expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            expect(document.activeElement).toEqual(button);
          });

          it('updates the displayed value', () => {
            const { container } = render(<DropDownWrapper options={options} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            fireEvent.keyDown(document.activeElement, { key: 'Enter' });
            expect(button).toHaveTextContent('Banana');
          });
        });

        describe('when the value is the current value', () => {
          it('closes the listbox without calling setValue', () => {
            const spy = jest.fn();
            const { container } = render(<DropDownWrapper options={options} value="Banana" setValue={spy} />);
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'Enter' });
            expect(spy).not.toHaveBeenCalled();
            expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            expect(document.activeElement).toEqual(button);
          });
        });
      });

      describe('when blurring the listbox', () => {
        describe('when the value has changed', () => {
          it('calls setValue', async () => {
            const spy = jest.fn();
            const { container } = render((
              <>
                <DropDownWrapper options={options} setValue={spy} />
                <input />
              </>
            ));
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            container.querySelector('input').focus();
            await wait(() => {
              expect(spy).toHaveBeenCalledWith('Banana');
            });
          });

          it('closes the list box without removing the focus', async () => {
            const { container } = render((
              <>
                <DropDownWrapper options={options} />
                <input />
              </>
            ));
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            const input = container.querySelector('input');
            input.focus();
            await wait(() => {
              expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            });
            expect(document.activeElement).toEqual(input);
          });

          it('updates the displayed value', async () => {
            const { container } = render((
              <>
                <DropDownWrapper options={options} />
                <input />
              </>
            ));
            const button = container.querySelector('button');
            fireEvent.click(button);
            fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
            const input = container.querySelector('input');
            input.focus();
            await wait(() => {
              expect(button).toHaveTextContent('Banana');
            });
          });
        });

        describe('when the value has not changed', () => {
          it('closes the listbox without calling setValue', async () => {
            const spy = jest.fn();
            const { container } = render((
              <>
                <DropDownWrapper options={options} value="Banana" setValue={spy} />
                <input />
              </>
            ));
            const button = container.querySelector('button');
            fireEvent.click(button);
            const input = container.querySelector('input');
            input.focus();
            await wait(() => {
              expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
            });
            expect(spy).not.toHaveBeenCalled();
            expect(document.activeElement).toEqual(input);
          });
        });
      });
    });
  });

  describe('options as array of numbers', () => {
    const options = [1, 2, 3];

    it('renders a closed drop down', () => {
      const { container } = render(<DropDownWrapper options={options} />);
      expect(container).toMatchSnapshot();
      expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
    });

    it('triggers the setValue callback with the selected value', () => {
      const spy = jest.fn();
      const { container } = render(<DropDownWrapper options={options} setValue={spy} />);
      const button = container.querySelector('button');
      fireEvent.click(button);
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'Enter' });
      expect(spy).toHaveBeenCalledWith(2);
      expect(container.querySelector('[role=listbox]')).toHaveAttribute('hidden', '');
      expect(document.activeElement).toEqual(button);
    });
  });

  describe.skip('options as array of objects', () => {
    describe('label', () => {
      const options = [{ label: 'Apple' }, { label: 'Banana' }, { label: 'Orange' }];

      it('renders an select', () => {
        const { container } = render(
          <SelectWrapper options={options} />,
        );
        expect(container).toMatchSnapshot();
      });

      it('triggers the setValue callback with the selected value', () => {
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} setValue={spy} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '1' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Banana' });
      });

      it('updates when the value changes', () => {
        const { container } = render(<SelectWrapper options={options} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '1' } });
        expect(select).toHaveValue('1');
        expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('Banana');
      });
    });

    describe('value', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', value: 1 }, { label: 'foo', value: 2 }, { label: 'foo', value: 3 }];
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} value={2} setValue={spy} />);
        const select = container.querySelector('select');
        expect(select).toHaveValue('1');
        fireEvent.change(select, { target: { value: '2' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', value: 3 });
      });
    });

    describe('id', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', id: 1 }, { label: 'foo', id: 2 }, { label: 'foo', id: 3 }];
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} value={2} setValue={spy} />);
        const select = container.querySelector('select');
        expect(select).toHaveValue('1');
        fireEvent.change(select, { target: { value: '2' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', id: 3 });
      });
    });

    describe('disabled', () => {
      it('sets the disabled attribute', () => {
        const options = [{ label: 'foo', disabled: true }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).toHaveAttribute('disabled', '');
      });
    });

    describe('html', () => {
      it('sets attributes on the element', () => {
        const options = [{ label: 'foo', html: { 'data-foo': 'bar', className: 'class' } }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).toHaveAttribute('data-foo', 'bar');
        expect(option).toHaveAttribute('class', 'class');
      });
    });

    describe('group', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Orange', group: 'Citrus' },
        { label: 'Lemon', group: 'Citrus' },
        { label: 'Raspberry', group: 'Berry' },
        { label: 'Strawberry', group: 'Berry' },
      ];

      it('renders grouped options', () => {
        const { container } = render(<SelectWrapper options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('triggers the setValue callback with the selected value', () => {
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} setValue={spy} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '3' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Lemon', group: 'Citrus' });
      });

      it('updates when the value changes', () => {
        const { container } = render(<SelectWrapper options={options} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '3' } });
        expect(select).toHaveValue('3');
        expect(select.querySelector('optgroup:first-of-type > option:nth-child(2)')).toHaveTextContent('Lemon');
      });
    });

    describe('other attributes', () => {
      it('does not render them', () => {
        const options = [{ label: 'foo', 'data-foo': 'bar' }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).not.toHaveAttribute('data-foo', 'bar');
      });
    });
  });

  describe.skip('mapOption', () => {
    const options = [{ name: 'Apple' }, { name: 'Banana' }, { name: 'Orange' }];

    it('maps options', () => {
      const spy = jest.fn();
      const { container } = render(<SelectWrapper
        options={options}
        setValue={spy}
        mapOption={({ name }) => ({ label: name })}
      />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '2' } });
      expect(spy).toHaveBeenCalledWith({ name: 'Orange' });
    });

    it('selects a mapped option', () => {
      const { container } = render(<SelectWrapper
        options={options}
        value={{ name: 'Banana' }}
        mapOption={({ name }) => ({ label: name })}
      />);
      const select = container.querySelector('select');
      expect(select).toHaveValue('1');
      fireEvent.change(select, { target: { value: '2' } });
      expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('Banana');
    });
  });
});

describe.skip('blank', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('renders a blank option', () => {
    const { container } = render(<SelectWrapper options={options} blank="Please select…" />);
    const select = container.querySelector('select');
    expect(select.querySelector('option')).toHaveTextContent('Please select…');
    expect(select).toHaveValue('0');
  });

  it('renders with a selected value', () => {
    const { container } = render(<SelectWrapper options={options} blank="Please select…" value="Orange" />);
    const select = container.querySelector('select');
    expect(select).toHaveValue('3');
    expect(select.querySelector('option:nth-child(4)')).toHaveTextContent('Orange');
  });
});

describe.skip('OptionComponent', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('allows additional props to be added to all options', () => {
    const { container } = render(
      <SelectWrapper options={options} OptionComponent={{ 'data-foo': 'bar' }} />,
    );
    container.querySelectorAll('option').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });

  it('allows the option component to be replaced', () => {
    function TestOption({ props }) {
      return (
        <option
          data-foo="bar"
          {...props}
        />
      );
    }
    const { container } = render(
      <SelectWrapper options={options} OptionComponent={TestOption} />,
    );
    container.querySelectorAll('option').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe.skip('OptGroupComponent', () => {
  const options = [
    { label: 'Apple' },
    { label: 'Orange', group: 'Citrus' },
    { label: 'Lemon', group: 'Citrus' },
    { label: 'Raspberry', group: 'Berry' },
    { label: 'Strawberry', group: 'Berry' },
  ];

  it('allows additional props to be added to all optgroups', () => {
    const { container } = render(
      <SelectWrapper options={options} OptGroupComponent={{ 'data-foo': 'bar' }} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });

  it('allows the optgroup component to be replaced', () => {
    function TestOptGroup({ props }) {
      return (
        <optgroup
          data-foo="bar"
          {...props}
        />
      );
    }
    const { container } = render(
      <SelectWrapper options={options} OptGroupComponent={TestOptGroup} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe.skip('SelectComponent', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('allows additional props to be added to the select', () => {
    const { container } = render(
      <SelectWrapper options={options} SelectComponent={{ 'data-foo': 'bar' }} />,
    );
    expect(container.querySelector('select')).toHaveAttribute('data-foo', 'bar');
  });

  it('allows the option component to be replaced', () => {
    function TestSelect({ props }) {
      return (
        <datalist
          data-foo="bar"
          {...props}
        />
      );
    }
    const { container } = render(
      <SelectWrapper options={options} OptGroupComponent={TestSelect} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe.skip('additional props', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('includes them on the select', () => {
    const { container } = render(
      <SelectWrapper options={options} required data-foo="bar" />,
    );
    const select = container.querySelector('select');
    expect(select).toHaveAttribute('required', '');
    expect(select).toHaveAttribute('data-foo', 'bar');
  });
});
