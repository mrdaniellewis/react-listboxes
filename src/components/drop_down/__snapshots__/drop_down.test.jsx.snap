// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options as array of objects disabled sets the aria-disabled attribute 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-controls="id_listbox"
      aria-haspopup="listbox"
      class="dropdown__button"
      id="id"
      type="button"
    >
      Apple
    </button>
    <ul
      class="dropdown__listbox"
      hidden=""
      id="id_listbox"
      role="listbox"
      tabindex="-1"
    >
      <li
        class="dropdown__option"
        id="id_apple"
        role="option"
        tabindex="-1"
      >
        Apple
      </li>
      <li
        aria-disabled="true"
        class="dropdown__option"
        id="id_banana"
        role="option"
        tabindex="-1"
      >
        Banana
      </li>
    </ul>
  </div>
</div>
`;

exports[`options as array of objects group renders grouped options 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-controls="id_listbox"
      aria-haspopup="listbox"
      class="dropdown__button"
      id="id"
      type="button"
    >
      Apple
    </button>
    <ul
      class="dropdown__listbox"
      hidden=""
      id="id_listbox"
      role="listbox"
      tabindex="-1"
    >
      <li
        class="dropdown__option"
        id="id_apple"
        role="option"
        tabindex="-1"
      >
        Apple
      </li>
      <li
        aria-disabled="true"
        aria-label="Citrus"
        class="dropdown__group"
        id="group_citrus"
        role="group"
        tabindex="-1"
      >
        Citrus
      </li>
      <li
        class="dropdown__option dropdown__option--grouped"
        id="id_orange"
        role="option"
        tabindex="-1"
      >
        Orange
      </li>
      <li
        class="dropdown__option dropdown__option--grouped"
        id="id_lemon"
        role="option"
        tabindex="-1"
      >
        Lemon
      </li>
      <li
        aria-disabled="true"
        aria-label="Berry"
        class="dropdown__group"
        id="group_berry"
        role="group"
        tabindex="-1"
      >
        Berry
      </li>
      <li
        class="dropdown__option dropdown__option--grouped"
        id="id_raspberry"
        role="option"
        tabindex="-1"
      >
        Raspberry
      </li>
      <li
        class="dropdown__option dropdown__option--grouped"
        id="id_strawberry"
        role="option"
        tabindex="-1"
      >
        Strawberry
      </li>
    </ul>
  </div>
</div>
`;

exports[`options as array of objects label renders a closed drop down 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-controls="id_listbox"
      aria-haspopup="listbox"
      class="dropdown__button"
      id="id"
      type="button"
    >
      Apple
    </button>
    <ul
      class="dropdown__listbox"
      hidden=""
      id="id_listbox"
      role="listbox"
      tabindex="-1"
    >
      <li
        class="dropdown__option"
        id="id_apple"
        role="option"
        tabindex="-1"
      >
        Apple
      </li>
      <li
        class="dropdown__option"
        id="id_banana"
        role="option"
        tabindex="-1"
      >
        Banana
      </li>
      <li
        class="dropdown__option"
        id="id_orange"
        role="option"
        tabindex="-1"
      >
        Orange
      </li>
    </ul>
  </div>
</div>
`;

exports[`options options as array of numbers renders a closed drop down 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-controls="id_listbox"
      aria-haspopup="listbox"
      class="dropdown__button"
      id="id"
      type="button"
    >
      1
    </button>
    <ul
      class="dropdown__listbox"
      hidden=""
      id="id_listbox"
      role="listbox"
      tabindex="-1"
    >
      <li
        class="dropdown__option"
        id="id_1"
        role="option"
        tabindex="-1"
      >
        1
      </li>
      <li
        class="dropdown__option"
        id="id_2"
        role="option"
        tabindex="-1"
      >
        2
      </li>
      <li
        class="dropdown__option"
        id="id_3"
        role="option"
        tabindex="-1"
      >
        3
      </li>
    </ul>
  </div>
</div>
`;

exports[`options options as array of strings renders a closed drop down 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-controls="id_listbox"
      aria-haspopup="listbox"
      class="dropdown__button"
      id="id"
      type="button"
    >
      Apple
    </button>
    <ul
      class="dropdown__listbox"
      hidden=""
      id="id_listbox"
      role="listbox"
      tabindex="-1"
    >
      <li
        class="dropdown__option"
        id="id_apple"
        role="option"
        tabindex="-1"
      >
        Apple
      </li>
      <li
        class="dropdown__option"
        id="id_banana"
        role="option"
        tabindex="-1"
      >
        Banana
      </li>
      <li
        class="dropdown__option"
        id="id_orange"
        role="option"
        tabindex="-1"
      >
        Orange
      </li>
    </ul>
  </div>
</div>
`;
